### PR TITLE
[Merged by Bors] - feat: integration by parts for line derivatives and Frechet derivatives

### PR DESCRIPTION
**Pull request by sgouezel**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by ADedecker** [comment]
@mcdoll Didn't you work on something similar? If so, how does your version compare to this?

**Comment by mcdoll** [comment]
Yes, but this looks way better than what I did.

**Comment by mcdoll** [comment]
LGTM.
My only concern is that the assumptions are not really what one usually checks, aka I would like to have variants which assume that the derivatives are integrable and the functions itself are continuous and vanishes at infinity. If you want to add these in a later PR that is totally fine by me.

**Comment by sgouezel** [comment]
Indeed it's not the most standard set of assumptions. It's not directly comparable (my assumptions don't imply that the function tends to zero at infinity!). The assumptions I use are the ones that are minimal for the applications to Fourier transform. Let me add a comment with a TODO about that.

File: Mathlib/Analysis/Calculus/LineDeriv/IntegrationByParts.lean [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+    (hfg' : Integrable (fun x ↦ B (f x) (g' x)) (μ.prod volume))
+    (hfg : Integrable (fun x ↦ B (f x) (g x)) (μ.prod volume))
+    (hf : ∀ x, HasLineDerivAt ℝ f (f' x) x (0, 1)) (hg : ∀ x, HasLineDerivAt ℝ g (g' x) x (0, 1)) :
+    ∫ x, B (f x) (g' x) ∂(μ.prod volume) = - ∫ x, B (f' x) (g x) ∂(μ.prod volume) := calc
+  ∫ x, B (f x) (g' x) ∂(μ.prod volume)
+    = ∫ x, (∫ t, B (f (x, t)) (g' (x, t))) ∂μ := integral_prod _ hfg'
+  _ = ∫ x, (- ∫ t, B (f' (x, t)) (g (x, t))) ∂μ := by
+    apply integral_congr_ae
+    filter_upwards [hf'g.prod_right_ae, hfg'.prod_right_ae, hfg.prod_right_ae]
+      with x hf'gx hfg'x  hfgx
```
Comment (Ruben-VandeVelde) [inline Mathlib/Analysis/Calculus/LineDeriv/IntegrationByParts.lean:65]:
```suggestion
      with x hf'gx hfg'x hfgx
```

**Review by Ruben-VandeVelde — APPROVED** [review]

**Comment by Ruben-VandeVelde** [comment]
maintainer merge

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**